### PR TITLE
Simplified how ignored tokens are reported

### PR DESCRIPTION
**Pull request by superbobry**
I haven't yet wrapped my head about the interaction between events and shift/reduce, so the PR is only an RFC at this point.

### CODE REVIEW

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp [hunk 1]
```diff
@@ -294,11 +286,11 @@ ${end-}
 				}
 			}
 ${if self->ignoredReportTokens()-}
```
Comment (inspirer) [inline tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp:288]: I think flushing the pending tokens here is a bug. If the following recoverFromError recovers without shifting any tokens, and then we reduce `SyntaxProblem: error -> SyntaxProblem;`, this SyntaxProblem should be reported before any subsequent comments.

We should flush p.pending in front of every fetchNext and when we put noToken into p.next.symbol.

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp [hunk 2]
```diff
@@ -390,7 +378,7 @@ func (p *Parser) skipBrokenCode(lexer *Lexer, stack []stackEntry, canRecover fun
 		}
 ${call reportConsumedNext-}
 		e = p.next.endoffset
-		p.fetchNext(lexer, stack${if self->ignoredReportTokens()}, nil${end})
+		p.fetchNext(lexer, stack)
```
Comment (inspirer) [inline tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp:381]: Invariant: before calling fetchNext we should flush all pending tokens.

In most places p.pending is empty because of p.next.symbol == noToken, so we don't need to do anything. Here we should flush the previous batch _before_ reportConsumedNext. How about we extract "if len(p.pending) > 0 { for {...} ... }" into a flushPending template and call it from here as well?

File: tm-tool/textmapper/tests/org/textmapper/tool/bootstrap/bison/bison1.y [hunk 3]
```diff
@@ -20,101 +20,6 @@
 
 %%
 
-input :
```
Comment (inspirer) [inline tm-tool/textmapper/tests/org/textmapper/tool/bootstrap/bison/bison1.y:23]: unrelated change?

**Review by inspirer — COMMENTED** [review]
Thank you for doing this!

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp [hunk 4]
```diff
@@ -82,6 +87,9 @@ ${end-}
 ${if self->eventBased()-}
 	p.listener = l
 ${end-}
+${if self->ignoredReportTokens()-}
+	p.pending = make([]symbol, 0, startTokenBufferSize)
```
[the remark this answers is not in the document]
Comment (superbobry) [reply tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp:91]: Good idea! Done.

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp [hunk 5]
```diff
@@ -158,7 +166,7 @@ ${end}
 ${template parseFunc-}
 func (p *Parser) parse(${call contextParam}start, end ${self->stateType()}, lexer *Lexer) ${if !self->hasAssocValuesForInput()}error${else}(${self->valueType()}, error)${end} {
 ${if self->ignoredReportTokens()-}
-	ignoredTokens := make([]symbol, 0, startTokenBufferSize) // to be reported with the next shift
+	p.pending = p.pending[:0]
```
[the remark this answers is not in the document]
Comment (superbobry) [reply tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp:169]: Done.

File: tm-tool/textmapper/tests/org/textmapper/tool/bootstrap/bison/bison1.y (hunk 3 above)
Comment (superbobry) [reply tm-tool/textmapper/tests/org/textmapper/tool/bootstrap/bison/bison1.y:23]: It is unrelated, but it does occur whenever I run `./gen.sh`. Reverted for now.

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp (hunk 2 above)
Comment (superbobry) [reply tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp:375]: I like the idea of adding a template! I didn't realize it's an option. Done.

File: tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp (hunk 1 above)
Comment (superbobry) [reply tm-tool/textmapper/resources/org/textmapper/tool/templates/go_parser.ltp]: Done.
